### ISSUE TITLE
Container resize column recalculation

### ISSUE THREAD

**Issue by deeg** [swimlane/angular2-data-table#193]
**I'm submitting a ...**  (check one with "x")

```
[ ] bug report => search github for a similar issue or PR before submitting
[ x ] feature request
[ ] support request => Please do not submit support request here
```

**Current behavior**

It seems from [this line](https://github.com/swimlane/angular2-data-table/blob/90ad41d56ec58522b7793aaa6d1f9265ff7ed024/src/components/datatable.component.ts#L198) that the recalculation of column widths only happen on window resize event.

I have a container for a table which gets smaller if you open a filter section next to it. This does not fire that event and the recalculation does not happen.

**Expected behavior**

I would ask that we either bind to a resize event of the data table itself if that is possible, or give access to the [adjustSizes](https://github.com/swimlane/angular2-data-table/blob/90ad41d56ec58522b7793aaa6d1f9265ff7ed024/src/components/datatable.component.ts#L165) function so we can call it programmatically to re calculate the column widths.

**Reproduction of the problem**
http://plnkr.co/edit/wLIdtlUA0ZPrJL7nnpec?p=preview

Notice when clicking smaller/larger buttons to resize the container the columns don't recalculate. If you change the size of the window it will recalculate.

**What is the motivation / use case for changing the behavior?**
Being able to programmatically change the container size of the table, and have the table still render correctly.
- **Table version:** 0.7.x
  0.9.3
- **Angular version:** 2.0.x
  2.0.2
- **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
  Checked Chrome and FF both had this issue.
- **Language:** [all | TypeScript X.X | ES6/7 | ES5]
  TS 2.0.2

I am happy to work on a PR for this if it is something you are open to, but wanted to get feedback first.

**Comment by amcdnl** [comment]
Probably need something like this - http://www.backalleycoder.com/2013/03/18/cross-browser-event-based-element-resize-detection/

I would rather not expose the internal method.

**Comment by deeg** [comment]
Okay thanks, will look into that solution.

**Comment by ghetolay** [comment]
Angular is already doing a throttle so why not just do something like that : 

```
ngOnCheck() {
   if( this.element.offsetWidth != this.width || this.element.offsetHeight != this.height ) {
     this.width = this.element.offsetWidth;
     this.height = this.element.offsetHeight;
     doResize();
  }
}
```

You'll just have to mark the datatable for check.

**Comment by amcdnl** [comment]
@ghetolay suggestion is proper way.

**Comment by jinder** [comment]
@amcdnl did this suggestion make it into the component?  Or is there now a way to manually re-render the table on resize?

**Comment by deeg** [comment]
You can now control when to call recalculate on the table, so here is what I have working to recalculate the columns when the table container size changes:

```ts
  @ViewChild('tableWrapper') tableWrapper;
  @ViewChild(DatatableComponent) table: DatatableComponent;

  ngAfterViewChecked() {
    // Check if the table size has changed,
    if (this.table && this.table.recalculate && (this.tableWrapper.nativeElement.clientWidth !== this.currentComponentWidth)) {
      this.currentComponentWidth = this.tableWrapper.nativeElement.clientWidth;
      this.table.recalculate();
      this.changeDetectorRef.detectChanges();
    }
  }
```

```html
<div class="table-wrapper" #tableWrapper>
    <ngx-datatable #table ...></ngx-datatable>
</div>
```

**Comment by jinder** [comment]
@deeg Exactly what I was looking for - thank you.

**Comment by enough1987** [comment]
@deeg I did like in your exemple, but i have error 
![image](https://cloud.githubusercontent.com/assets/9271733/25276403/c17429ee-26a2-11e7-82e8-058808d3bcee.png)

```

  @ViewChild('tableWrapper') tableWrapper;
  @ViewChild(DatatableComponent) table: DatatableComponent;
  private currentComponentWidth;

  ngAfterViewChecked() {
    // Check if the table size has changed,
    if (this.table && this.table.recalculate && (this.tableWrapper.nativeElement.clientWidth !== this.currentComponentWidth)) {
      this.currentComponentWidth = this.tableWrapper.nativeElement.clientWidth;
      this.table.recalculate();
      this.changeDetectorRef.detectChanges();
      console.log( ' +++++++++++++++++++++++++++++++++++++++ ' );
    }
  }

```

```

<div #tableWrapper
     [attr.id]="wrapperId" 
     class="table-wrapper {{typeOfList}}" >

<!--     <div *ngIf="rows && rows.length"  >-->

     <ngx-datatable 
        #table
        class="material"
        [rows]="rows"
        [columns]="columns"
        [selected]="selected"
        [selectionType]="'multi'"
        [columnMode]="'force'"
        [headerHeight]="24"
        [footerHeight]="0"
        [rowHeight]="40"
        [scrollbarV]="true"
        [scrollbarH]="false"
        [messages]="{emptyMessage: 'No items have been added to the collection'}"
        (select)='onSelect($event)' 
        (page)='onPage($event)' >
```

**Comment by dylandrollut** [comment]
@amcdnl @deeg I'm having issues with getting multiple tables on the same page to resize. Is there anyway to target **ALL** tables on a page?

**Comment by dylandrollut** [comment]
A bit of a hacky solution that I found to resize all tables on the same page is to generate a window resize event in the type script file and tie it to whatever you use to trigger the page change.

**Comment by starcoda** [comment]
I used the approach proposed by @deeg, which worked well, but decided to wrap it in a directive:

```
import { Directive, AfterContentChecked } from '@angular/core';
import { DatatableComponent } from "@swimlane/ngx-datatable";

@Directive({ selector: '[resize-watcher]' })
export class ResizeWatcherDirective implements AfterContentChecked {

    constructor(private table: DatatableComponent) { }

    private latestWidth: number;

    ngAfterContentChecked() {
        if (this.table && this.table.element.clientWidth !== this.latestWidth) {
            this.latestWidth = this.table.element.clientWidth;
            this.table.recalculate();
        }
    }

}
```
Then you can just add a resize-watcher attribute to the table you want to recalculate.

**Comment by nikotincheg** [comment]
@starcoda I created NgxResizeWatcherDirective using your code and recalculate method was called, but my table didn't resized until I clicked on it(( I tries different solutions but looks like table.recalculate(); doesn't work for me....I really don't understand why...

My table:

<ngx-datatable ngx-resize-watcher
  class="my-table"
  [limit]="4"
  [columnMode]="'force'"
  [headerHeight]="40"
  [footerHeight]="40"
  [rowHeight]="'auto'"
  [loadingIndicator]="loading"
  [rows]="rows"
  [columns]="columns"
  [reorderable]="false">
</ngx-datatable>

**Comment by katestearns-suplari** [comment]
I also encountered a similar issue, did you find a solution?

**Comment by nikotincheg** [comment]
No, I didn't(

**Comment by starcoda** [comment]
@nikotincheg, @katestearns-suplari 

I updated my version to latest (I had 9.3.1 before) and started having the same issue. The recalculation was being called but the table wasn't redrawn. I resolved the issue by forcing change detection like so: 

```
import { Directive, AfterContentChecked, ChangeDetectorRef } from '@angular/core';
import { DatatableComponent } from "@swimlane/ngx-datatable";

@Directive({ selector: '[ngx-resize-watcher]' })
export class NgxResizeWatcherDirective implements AfterContentChecked {

    constructor(private table: DatatableComponent, public ref: ChangeDetectorRef) { }

    private latestWidth: number;

    ngAfterContentChecked() {
        if (this.table && this.table.element.clientWidth !== this.latestWidth) {
            this.latestWidth = this.table.element.clientWidth;
            this.table.recalculate();
            this.ref.markForCheck();
        }
    }

}
```

Hope this helps.

**Comment by starcoda** [comment]
Should probably be .detectChanges() rather than .markForCheck() actually. I'm new to Angular, so just figuring these things out. Either kicks off the refresh though.

**Comment by starcoda** [comment]
One more update to my previous comments. Having done a bit more digging and a bit more work with ngx-datatables, I have realized from the change log that everything was changed internally to use OnPush change detection in version 10.0.4 to improve performance. This however means that some of the changes that were previously triggering a redraw are no longer doing so. I have found other issues such as my [rowClass] not being evaluated and have therefore downgraded to version 10.0.3 for now until such time that I can put some effort into figuring out how to make my app work with the latest performance improvements.

**Comment by cstaub** [comment]
having the same issue with 10.0.3: recalculate() is executed but update is only performed after mouseover or mouseclick on datatable. Any solution?

**Comment by nextremos** [comment]
Same effect here with version 10.2.2 - update is only performed after mouseover or mouseclick on datatable. Any solution?

**Comment by stas-kh** [comment]
Guys, any update on this?

Really important thing to have it fixed.

**Comment by wizarrc** [comment]
@stas-kh I'm new to this thread, and looks like it's closed and problems are reoccurring. I can look into this for you if it's possible for you to give me a repro on say plunker.

You fork this http://plnkr.co/edit/YiKBIP1GjpbNOB1K9S28?p=preview 
I made it for another issue I was having. I know the one provided doesn't apply styles properly

**Comment by stas-kh** [comment]
@wizarrc Please take a look at the following:
http://plnkr.co/edit/gfr9IDIAdXEy9Am6B5PB?p=preview

Try to collapse, then expand sidebar and move your mouse over the table. The table changes its size only after some action like mousemove or scrolling.

**Comment by wizarrc** [comment]
@stas-kh thanks

**Comment by wizarrc** [comment]
I believe this happens now that the internal table is using `OnPush` change detection. Every column header cell and cell gets it's width and offset deep in the implementation, and yet it's only checked after an event that occurs in the actual cell (causing invalidation), in this case looks to be `mouseenter`. A simple table resize isn't enough to propagate checks deep inside the table structure.

**Comment by wizarrc** [comment]
@amcdnl I believe previous methods to make this work no longer applies.

**Comment by vosecek** [comment]
I can confirm the problem, can be patched by calling window.dispatchEvent(new Event('resize')); after recalculate() method.

setTimeout() is used due too sidebar hiding animation
```
setTimeout(() => {
   this.clientList.recalculate();
   window.dispatchEvent(new Event('resize'));
}, 250);
```

**Comment by ItsEcholot** [comment]
@vosecek This approach works, thanks! The timeout here is crucial if you have any kind of animation going on

**Comment by qinghuanchang** [comment]
一群辣鸡

**Comment by manijak** [comment]
This needs to be fixed asap!

**Comment by ionut95** [comment]
ngAfterViewChecked() {
    if (this.table && this.table.recalculate && (this.tableWrapper.nativeElement.clientWidth !== this.currentComponentWidth)) {
      this.currentComponentWidth = this.tableWrapper.nativeElement.clientWidth;
      this.table.recalculate();
      window.dispatchEvent(new Event('resize'));
    }
  }

**Comment by deeg** [comment]
@manijak your comment is no help here. Please remember the people working on this library are doing so voluntarily in their spare time.

Commenting on closed issues with no suggestions or fixes is not constructive. If it needs to be fixed asap, try putting out a PR with a fix so it can be merged. Please try to be more understanding in the future.

**Comment by manijak** [comment]
I agree @deeg , comment removed

**Comment by wizarrc** [comment]
Some hope of getting this fixed without requiring the user to do something on their end. I discovered a nifty library while fixing another issue:

https://github.com/marcj/css-element-queries

https://github.com/swimlane/ngx-datatable/pull/1033#issuecomment-334808175

> FYI, I did learn something while looking at the project mentioned above. The visibility directive is primarily used to refresh the table, similar to a windows resize. I thought the above project could essentially replace the visibility directive altogether. That's actually false. The `css-element-queries` project only works when the element is visible. But by having an event that can trigger when it's visible and disable once it's hidden again, this actually would fix the #193 without a user-required workaround or incurring a perf hit. I tried it out, and it looks like it can replace the `@HostListener('window:resize')` event as well. Since the window resize event can unnecessarily trigger refreshes even when the table isn't affected, it actually is more efficient.

Demo of issue 193 that `css-element-queries` can fix efficiently:
![sidebar](https://user-images.githubusercontent.com/8039397/31288636-bfe4109c-aa93-11e7-87ed-577b25aa5710.gif)

**Comment by szanellato93** [comment]
I tried all your workaround but i only goal a right resize after a mouseover. This is why angular doesn't listen my ref.detectChanges(). I tried with ChangeDetectorRef and with window.dispatchEvent(new Event('resize')); but nothing work. If i resize the window with mouse the table response correctly.

**Comment by jfox459** [comment]
I had to call `recalculateColumns()` instead of `recalculate()` for it to work completely.

**Comment by Hibbins** [comment]
I couldn't get it to work with any of the above solutions, still only adjusts when you mouseover. Any news on this ?

**Comment by szanellato93** [comment]
this problem still remain in the last 11.1.2 version? I can't go over the 10.4.0

**Comment by meeroslav** [comment]
@szanellato93 

`this.myTable.recalculate` calls `recalculateDims` and `recalculateColumns` so calling them again has no effect.

One thing that worked for me, was to feed new copy of columns to the `recalculate` method.
There is something off with recalculation function, so after initial calculation it'll remember the last widths, thus influencing every further calculation. That's why you have to feed it with clean TableColumns every time.

```ts
@HostListener('window:resize', ['$event'])
onResize(event) {
  this.fixBugColumnResize();
}

private get columns(): TableColumn[] {
  return [ /** your columns data here */ ];
}

fixBugColumnResize() {
  if (this.myTable && this.myTable.recalculate && (this.tableWrapper.nativeElement.clientWidth !== this.currentComponentWidth)) {
    this.currentComponentWidth = this.tableWrapper.nativeElement.clientWidth;
    this.myTable.recalculateColumns(this.columns);
  }
}
```

As soon as I have time, I'll create PR for calculation method, but until then you can use this trick.

**Comment by Ayoubiedas** [comment]
Refreshing data works for me
this.rows = [...this.rows];

**Comment by holuwaseun** [comment]
Yup @Ayoubiedas, this method works just fine. Spent hours trying to figure this out. Tried to trigger windows resize and even tried the table recalculate method.

My own problem was that I had a table within a sticky state and I needed to keep the data within the table updated since sticky states do not reload when you go back to them. I was successfully updating the data using socket, but whenever I redirected back to the sticky state, the columns of the table would have resized and the presentation was messed up. So what I did was to monitor route change and do a data refresh whenever the route changed to the sticky state.

Hope this helps someone.

Cheers

**Comment by BBaysinger** [comment]
@Ayoubiedas, @holuwaseun, That didn't work for me. I had to do this:

```
      setTimeout(() => {
        if (this.pager.group === 1) {
          this.table.bodyComponent.scroller.parentElement.scrollTop = 1;
          setTimeout(() => {
            this.table.bodyComponent.scroller.parentElement.scrollTop = 0;
          });
        }
      });
```

**Comment by mavivone426** [comment]
@BBaysinger can you provide a little more code to see how you implemented the setTimeout function?  would be most appreciated as I cannot solve this issue and is beyond frustrating...  Thanks again!

**Comment by BBaysinger** [comment]
@mavivone426, the rest of the code is pretty vast, but basically all I'm doing is waiting until after the rows are set.

**Comment by wes-dgo** [comment]
Hi, all! 
I think this issue needs to be reopened. (Maybe a new Issue cause different versions? ). 
Currently having the same issue: no update until scroll or window resize.   

Fixed it with a hacky redraw:
`onResize():void  { this.isDataLoaded = false;
 setTimeout( () => {
      this.isDataLoaded = true;
 }, 0);}`
and
`<div *ngIf="isDataLoaded"> ...rest of table... </div>`

**Comment by maximelafarie** [comment]
@wes-dgo fix inspired me and I tried some stuff...

Assuming you have this in your `Component`: 
```
public ready = false;
```
 and you also set this piece of code in the same `Component`:
```
ngAfterViewInit() {
  this.ready = true;
}
```

And you finally add this to the `Component`'s template:
```
<div class="my-container"> <!-- ngx-datatable shall take full width of this element  -->
  <ngx-datatable 
          *ngIf="ready"
          ...
          rest of code
          ...>
      ...
  </ngx-datatable>
</div>
```
It should work! 🤓

It works perfectly for me. Given that `ngx-datatable` wont appear until `ready` is `true`, `ngAfterViewInit()` will make it waiting the DOM is fully rendered.

So when `ready` is `true`, the container is already in the DOM and `ngx-datatable` has parent's width reference.

**Comment by vozgorkov** [comment]
It doesn't work with any of the above solutions(only mouseover etc), is somebody fix it?

**Comment by jblza** [comment]
Doesn't work with any of these solutions for me

**Comment by GillesMontyne** [comment]
same issue, the changed value only gets displayed onHover...

**Comment by kim24** [comment]
same issue, only refresh if on mouse over.

**Comment by ArfanMirza** [comment]
Here what i did to fix "refresh if on mouse over" issue/bug/problem...
Just follow the @deeg  solution and add one line only,

```
@ViewChild('tableWrapper') tableWrapper;
  @ViewChild(DatatableComponent) table: DatatableComponent;
  private currentComponentWidth;

  ngAfterViewChecked() {
    // Check if the table size has changed,
    if (this.table && this.table.recalculate && (this.tableWrapper.nativeElement.clientWidth !== this.currentComponentWidth)) {
      this.currentComponentWidth = this.tableWrapper.nativeElement.clientWidth;
      this.table.recalculate();
      

> $('.datatable-body-cell-label').click();

      this.changeDetectorRef.detectChanges();
    }
  }
```

I hope this make ppl life easy... :100:

**Comment by colinbull** [comment]
I doubt this will fix everyones' issues. But one really simple _fix_ for me when the table was bound inside a `tab group` was to simply add an `*ngIf` on the table that looked at the rows variable. This seemed to delay the bounds computation enough so by the time the tab had transitioned, the correct client width was available. YMMV... 

           <mat-tab-group>
          <mat-tab label="Buy Trades">
            <ngx-datatable *ngIf="buyTrades"
              class="material"
              [columnMode]="'force'"
              [rows]="buyTrades"
              [columns]="tradeColumns"
              [headerHeight]="40"
              [rowHeight]="30"
              [scrollbarV]="true">
            </ngx-datatable>
          </mat-tab>
          <mat-tab label="Sell Trades">
            <ngx-datatable *ngIf="sellTrades"
              class="material"
              [columnMode]="'force'"
              [rows]="sellTrades"
              [columns]="tradeColumns"
              [headerHeight]="40"
              [rowHeight]="30"
              [scrollbarV]="true">
            </ngx-datatable>
          </mat-tab>
        </mat-tab-group>

**Comment by phrixothrixa** [comment]
Hello everyone! I had the same problem "refresh if on mouse over" with mat-sidebar today. 
I spent hours trying to figure this out. Tried to trigger windows resize and even tried the table recalculate method and with any of the above solutions it didn't work.
But I noticed that the table resizes when 'window:resize' event happened on browser. Than I add 
<code>
setTimeout(() => {
    window.dispatchEvent(new Event('resize'));
}, 250)
</code>
to 'onMouseLeave ' and 'onMouseEnter' functions of sidebar. And It works for me!

**Comment by ShinDarth** [comment]
same problem here

**Comment by andreyshedko** [comment]
None of the solutions working for me.

**Comment by ArfanMirza** [comment]
@andreyshedko  can give detail what you are facing issue ?

**Comment by qutayba** [comment]
I know many tried this but it worked for me. I noticed that the recalculation of the width triggers with every hover/click. So I trigger click event (in `ngOnInit`) on the label of the cells in rows and header:

`ngOnInit() {

    setTimeout(() => {
      $('.datatable-body-cell-label').click();
      $('.datatable-header-cell-label').click();
    }, 300);
  }`

Still pretty hacky and it should be fixed but it's temporary. I hope it will help anyone

**Comment by Ardeshir81** [comment]
@Ayoubiedas 's solution (refreshing data) worked for me too! Thanks!

**Comment by radoslavpetranov** [comment]
Why is this issue closed?

**Comment by johnpozy** [comment]
In my case, the datatable is inside ngx-bootstrap tab. Each time user navigate between tab, the datatable columns will get shrink the below code fix my problem:

`const event = window.document.createEvent('UIEvents');setTimeout(() => {
      event.initUIEvent('resize', true, false, window, 0);
      window.dispatchEvent(event);
    }, 0);`

**Comment by ankitsalu** [comment]
if (navigator.userAgent.indexOf('MSIE') !== -1 || navigator.appVersion.indexOf('Trident/') > 0) { // IE-11 hack
            var evt = document.createEvent('UIEvents');
            evt.initUIEvent('resize', true, false, window, 0);
            window.dispatchEvent(evt);
        } else { // work in all browser
            window.dispatchEvent(new Event('resize'));
        }

**Comment by sivanargana** [comment]
recalculate is still not working with sidebar toggle

**Comment by neiljeffries** [comment]
After days of fighting with this I finally found a solution that worked for me. **The key was setting this.rows inside of a timeout.** Hope this helps someone. 

    this.myService.myData$.subscribe((data) => {
      setTimeout(() => {
        this.rows = [...data];
      });
    });

**Comment by sagar-wadke** [comment]
https://stackblitz.com/edit/angular-ngx-datatable-resize-issue

This is the solution for <ngx-datatable> resize issue.

**Comment by flue89** [comment]
create model: Ngx-datatable-resizer.ts

```
export class NgxDatatableResizer {

  name: string;
  grow: number;

  constructor(name: string, grow: number) {
    this.name = name;
    this.grow = grow;
  }
}

```

create helper funcion: table-resize2.ts

```
import {NgxDatatableResizer} from "../../../model/Ngx-datatable-resizer";

/**
 * @author: Mateusz Zastawny
 */

export class TableResize2 {

  private tableId: string;
  private columns: Array<NgxDatatableResizer>;
  private columnsMap = {};

  private sum: number = 0;

  constructor(tableId: string, columns: Array<NgxDatatableResizer>) {
    this.tableId = tableId;
    this.columns = columns;

    this.setMap();
  }

  private setMap(): void {
    this.columns.forEach(column => {
      this.columnsMap[column.name] = column.grow;
      this.sum = Math.round((this.sum += column.grow) * 100) / 100;
    });
  }

  public width(name: string): number {
    const maxWidth = Number(window.getComputedStyle(document.getElementById(this.tableId)).width.slice(0, -2));
    return (Math.round((((this.columnsMap[name] * 100) / this.sum) / 100) * 100) / 100) * maxWidth;
  }
}
```

use in .ts component:

```
//
public tableIdNameResizer: TableResize2 = new TableResize2('tableIdName',
    [
      new NgxDatatableResizer('name1', 0.3),
      new NgxDatatableResizer('name2', 0.4),
      new NgxDatatableResizer('name3', 0.87)
    ]);
//

```

use in .html component:

```
<ngx-datatable
        id="tableIdName"
        [columnMode]="'force'"
        >
        <ngx-datatable-column [minWidth]="30" [width]="tableIdNameResizer.width('name1')" [canAutoResize]="true" name="NAME1" prop="name1">
          <!--..-->
        </ngx-datatable-column>
        <ngx-datatable-column [minWidth]="30" [width]="tableIdNameResizer.width('name2')" [canAutoResize]="true" name="NAME2" prop="name2">
          <!--..-->
        </ngx-datatable-column>
        <!--..-->
      </ngx-datatable>
```

**Comment by erikhofer** [comment]
Can we re-open this? The problem is still present. I'm on the latest version working with Angular 8 (16.1.1). The columns are not only broken after resize but also initially when I use a frozen right column.

There is a PR that links to this issue which was recently closed: #1040 
It looks promising, maybe we can build on this?

This is the only workaround I got wo work. It is, however, sometimes lagging behind as much as a second.
```
  @ViewChild('wrapper', { static: true })
  wrapper: ElementRef

  lastWrapperWidth = 0

  ngAfterViewChecked() {
    if (this.wrapper.nativeElement.clientWidth !== this.lastWrapperWidth) {
      this.lastWrapperWidth = this.wrapper.nativeElement.clientWidth
      setTimeout(() => {
        window.dispatchEvent(new Event('resize'))
      })
    }
  }
```

**Comment by meenu1** [comment]
@ViewChild("table") table: DatatableComponent;

@HostListener('window:resize', ['$event'])
onResize(event) {
this.table.recalculate();
this.table.recalculateColumns();
}

This works for me.

**Comment by porlov** [comment]
Any updates on this issue?
I didn't manage to work any solution suggested above.

Thanks!

**Comment by porlov** [comment]
I found the solution. If someone will face with the same issue:
```js
@HostListener('window:resize', ['$event'])
onResize() {
  const redrawTable = () => {
    this.currentComponentWidth = this.tableWrapper.nativeElement.clientWidth;
    this.table.recalculate();

    setTimeout(() => {
      if (this.tableWrapper.nativeElement.clientWidth !== this.currentComponentWidth) {
        window.requestAnimationFrame(redrawTable);
      }
    }, 20);
  };

  window.requestAnimationFrame(redrawTable);
}
```

**Comment by MRROOX** [comment]
My Solution:

1). Install   angular-resize-event . Doc: https://www.npmjs.com/package/angular-resize-event
2). Add AngularResizedEventModule on AppModule
3). Inside the component where is the ngx-datatable add this method:


MyComponent.ts
```
import { ResizedEvent } from 'angular-resize-event';

.....

    public onResized($event: ResizedEvent): void {
        // console.log('ResizedEvent ' + $event);
        window.dispatchEvent(new Event('resize'));
    }
```
MyComponent.html


```
<ngx-datatable   (resized)="onResized($event)"> class="material" [rows]="rows"
  [columnMode]="ColumnMode.force" [columns]="columns" [headerHeight]="50"
  [footerHeight]="50" rowHeight="auto" [limit]="5" [selected]="selected"
  [selectionType]="SelectionType.single" (activate)="onActivate($event)"
  (select)="onSelect($event)">
</ngx-datatable>

```


This Work for Me.

**Comment by lapitzasclub** [comment]
For me for ngx-datatables with scroll... I have to made this modification in the @starcoda 's code:

```javascript
@Directive({ selector: '[ngx-resize-watcher]' })
export class NgxResizeWatcherDirective implements AfterContentChecked {
    private latestWidth: number;

    constructor(private table: DatatableComponent, public ref: ChangeDetectorRef) { }

    ngAfterContentChecked() {
        if (this.table && this.table.element.clientWidth !== this.latestWidth) {
            this.latestWidth = this.table.element.clientWidth;
            this.table.recalculate();
            this.ref.markForCheck();
            window.dispatchEvent(new Event('resize'));
        }
    }
}
```

**Comment by M4G0-AX** [comment]
My Solution:

I created this code with the solution that @MRROOX gave.

1). Install css-element-queries. Doc: https://www.npmjs.com/package/css-element-queries
2). Create this directive

`import { Directive, ElementRef, OnDestroy, OnInit } from '@angular/core';
import { ResizeSensor } from 'css-element-queries';

@Directive({
  selector: '[resize]',
})
export class ResizeDirective implements OnInit, OnDestroy {
  private resizeSensor: ResizeSensor;

  constructor(private readonly element: ElementRef) {}

  ngOnInit(): void {
    if (ResizeSensor) {
      this.resizeSensor = new ResizeSensor(this.element.nativeElement, () =>
        window.dispatchEvent(new Event('resize'))
      );
    }
  }

  ngOnDestroy(): void {
    if (this.resizeSensor) {
      this.resizeSensor.detach();
    }
  }
}`

**Comment by ezzabuzaid** [comment]
Most of the mentioned solutions worked for me, but still when resizing it freezz for a moment before resizing.

What really worked is using the resize observer as following  

```typescript
import { AfterViewInit, Directive, ElementRef, OnDestroy } from '@angular/core';
import { DatatableComponent } from "@swimlane/ngx-datatable";

@Directive({
    selector: '[ngx-resize-watcher]'
})
export class NgxResizeWatcherDirective implements AfterViewInit, OnDestroy {
    resizeObserver: ResizeObserver;
    constructor(
        private elementRef: ElementRef<HTMLElement>,
        private table: DatatableComponent,
    ) { }
    ngAfterViewInit(): void {
        this.resizeObserver = new ResizeObserver(entries => {
            this.table.recalculate();
        });
        this.resizeObserver.observe(this.elementRef.nativeElement);
    }

    ngOnDestroy(): void {
        this.resizeObserver.unobserve(this.elementRef.nativeElement);
    }

}
```

**Comment by alexandis** [comment]
> [columnMode]="'force'"

Tried this, but did not help: `this.table.recalculate()` does nothing and the table still spreads beyond my content area... I thought that probably I just need to set proper ngStyle width for this table, calculating viewscreen area width minus `mat-sidenav` width

**Comment by alexandis** [comment]
> For me for ngx-datatables with scroll... I have to made this modification in the @starcoda 's code:

Does not help - my table still goes beyond my material design content area width

**Comment by alexandis** [comment]
> you can now control when to call recalculate on the table, so here is what I have working to recalculate the columns when the table container size change

I debugged this for my case (material design with the grid in content area, left menu is collapsible): and - probably - the sizes are calculated BEFORE left material menu is created / shown. So the grid size is calculated incorrectly and then - probably - wrapper width is adjusted accordingly. As result, grid does not fit content area and all parent containers too. Disaster. But I did not have this kind of problem with old table component (it used `table` though).

**Comment by rathodrajnesh** [comment]
None of solution is working with me too. having version     "@swimlane/ngx-datatable": "^20.0.0",

**Comment by alexandis** [comment]
I made it using `MutationObserver` which tracks style change for left menu and sets `DatatableComponent` width.

    @Directive({ selector: '[sidePanelToggle]' })
    export class SidePanelToggleDirective implements OnDestroy {
      private changes: MutationObserver;
      wasVisible: boolean | undefined;
    
      constructor(
        private elementRef: ElementRef,
        @Host() @Self() private tableComponent: DatatableComponent
      ) {
        this.changes = new MutationObserver((mutations: MutationRecord[]) => {
            mutations.forEach((mutation: MutationRecord) => {
                const sideNavElement = mutation.target as HTMLElement;
                const tableElement = this.elementRef.nativeElement as HTMLElement;
                const isVisible = sideNavElement?.style.visibility === 'visible';
                if (this.wasVisible === undefined || this.wasVisible && !isVisible || !this.wasVisible && isVisible) {
                  if (tableElement.style.width) {
                    tableElement.style.width = `${SidePanelToggleDirective.floatOrZero(tableElement.style.width) + (isVisible ? -sideNavElement.offsetWidth : sideNavElement.offsetWidth)}px`;
                    console.log('Already set before, now will be ', tableElement.style.width);
                  }
                  else {
                    tableElement.style.width = `${tableElement.offsetWidth + (isVisible ? -sideNavElement.offsetWidth : sideNavElement.offsetWidth)}px`;
                    console.log('Setting first time as ', tableElement.style.width);
                  }
                  this.tableComponent.recalculate();
                }
                this.wasVisible = isVisible;
            });
        });
    
        document.querySelectorAll('mat-sidenav[position="start"]').forEach(element => {
          this.changes.observe(element, { attributeFilter: ['style'] });
        });
      }
    
      private static floatOrZero(value: any) {
        const result = parseFloat(value);
        return isNaN(result) ? 0 : result;
      }
    
      ngOnDestroy(): void {
        this.changes.disconnect();
      }
    }

Change in markup is minimal:

    <ngx-datatable sidePanelToggle ...

**Comment by yrgbel** [comment]
For me this solution with using directive works fine:
```
import { Directive, OnDestroy, OnInit } from '@angular/core';

import { DatatableComponent } from "@swimlane/ngx-datatable";

@Directive({
    selector: '[resizeTableWatcher]'
})
export class ResizeTableWatcherDirective implements OnInit, OnDestroy {
    private resizeObserver: ResizeObserver;

    constructor(private table: DatatableComponent) { }

    ngOnInit(): void {
        this.resizeObserver = new ResizeObserver(() => window.dispatchEvent(new Event('resize')));
        this.resizeObserver.observe(this.table.element);
    }

    ngOnDestroy(): void {
         this.resizeObserver.unobserve(this.table.element);
    }
}
```
then in the html template:
`<ngx-datatable resizeTableWatcher ...> ... </ngx-datatable>`

I also tested it with ngAfterViewInit method instead of the ngOnInit method and it worked fine as well.

**Comment by Krishnaaugustine** [comment]
latestWidth: any;
  constructor(private activatedRoute: ActivatedRoute, private router: Router, private location: Location, public ref: ChangeDetectorRef,
    private _activetab: TabserviceService, private virtualTableService: VirtualTableService, private token: TokenStorage,
    private dataMatrixService: DataMatrixService, private formService: FormService, private reportService: GenerateReportService) { }
  ngAfterContentChecked(): void {
    this.recalculate()
  }
  recalculate() {
    if (this.dataGrid && this.dataGrid.element.clientWidth !== this.latestWidth) {
      this.latestWidth = this.dataGrid.element.clientWidth;
      setTimeout(() => {
        this.dataGrid.recalculate();
        window.dispatchEvent(new Event('resize'));
     }, 250);
     this.ref.markForCheck();
    }
  }
